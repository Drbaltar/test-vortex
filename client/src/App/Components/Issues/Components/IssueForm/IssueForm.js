import React, { Component } from 'react';

import SelectBox from '../../../shared-components/SelectBox';
import TextArea from '../../../shared-components/TextArea';
import ErrorMessage from '../../../shared-components/ErrorMessage';
import BodyCard from '../../../shared-components/BodyCard/BodyCard';

class IssueForm extends Component {

    constructor (props) {
        super(props);

        this.FormFooter = this.props.children;

        this.applicationCategories = [
            '',
            'Application Crash',
            'Entry/List Not Populating Information',
            'Incorrect Information/Description',
            'Incorrect Spelling/Terminology',
            'Server Error',
            'Request New Feature',
            'Recommend Application Change',
            'Other'
        ];
    
        this.questionCategories = [
            '',
            'Incorrect Question Information',
            'Outdated Information',
            'Incorrect Answer',
            'Other'
        ];

        if (props.data) {
            this.setFilledInitialState(props.data);
        } else {
            this.setBlankInitialState();
        }

        this.state = this.initialState;
    }

    setFilledInitialState = (data) => {
        this.initialState = {
            issueType: data.issue_type,
            issueCategory: data.issue_category,
            issueDescription: data.issue_description,
            questionID: data.question_id || '',
            inputValidity: {
                isIssueTypeValid: null,
                isIssueCategoryValid: null,
                isIssueDescriptionValid: null,
                isQuestionIDValid: null
            }
        };
    }

    setBlankInitialState = () => {
        this.initialState = {
            issueType: '',
            issueCategory: '',
            issueDescription: '',
            questionID: '',
            inputValidity: {
                isIssueTypeValid: null,
                isIssueCategoryValid: null,
                isIssueDescriptionValid: null,
                isQuestionIDValid: null
            }
        };
    }

    render() {
        return(
            <BodyCard title={this.props.title}>
                {this.getFormBody()}
                {this.getFormFooter()}
            </BodyCard>
        );
    }

    getFormBody = () => {
        return (
            <div id='issue-form-body'>
                {this.getIssueTypeField()}
                {this.getIssueCategoryField()}
                {this.getIssueDescriptionField()}
                {this.props.errorMessage ? this.getErrorMessage() : null}
            </div>
        );
    }

    getIssueTypeField = () => {
        return (
            <SelectBox label="Issue Type" id="issueType"
                options={['', 'Application Issue', 'Question Issue']}
                value={this.state.issueType}
                inputChange={(event) => this.handleInputChange(event)}
                isValid={this.state.inputValidity.isIssueTypeValid}
                errorMessage='You must select the issue type!'/>
        );
    }

    getIssueCategoryField = () => {
        return (
            <SelectBox label="Issue Category" id="issueCategory"
                options={this.getIssueCategories()}
                value={this.state.issueCategory}
                inputChange={(event) => this.handleInputChange(event)}
                isValid={this.state.inputValidity.isIssueCategoryValid}
                errorMessage='You must select the issue category!'/>
        );
    }

    getIssueCategories = () => {
        if (this.state.issueType === 'Application Issue') {
            return this.applicationCategories;
        } else if (this.state.issueType === 'Question Issue') {
            return this.questionCategories;
        } else {
            return [];
        }
    }

    getIssueDescriptionField = () => {
        return (
            <TextArea label="Issue Description" id="issueDescription" type="text" rows="4"
                value={this.state.issueDescription}
                inputChange={(event) => this.handleInputChange(event)}
                isValid={this.state.inputValidity.isIssueDescriptionValid}
                errorMessage='The issue description must be at least 10 characters!'/>
        );
    }

    getErrorMessage = () => {
        return <ErrorMessage message={this.props.errorMessage}/>;
    }

    getFormFooter = () => {
        return React.cloneElement(this.FormFooter, {
            clickHandler: (e) => this.handleClickEvent(e)
        });
    }

    handleInputChange = (event) => {
        const {target: { id, value}} = event;

        if (id === 'issueType') {
            this.setState({[id]: value, issueCategory: ''});
        } else {
            this.setState({[id]: value});
        }
    }

    handleClickEvent = (event) => {
        event.preventDefault();

        if (event.target.id === 'clearAllButton') {
            this.setState(this.initialState);
            this.props.clearErrorMessage();
        } else if (this.isAllInputValid()) {
            this.props.submit(event.target.id, this.getDataToSubmit());
        }
    }

    isAllInputValid = () => {
        let inputValidity = this.getInputValidity();
        this.setState({ inputValidity });

        for (const entry in inputValidity) {
            if (!inputValidity[entry]) {
                return false;
            }
        }

        return true;
    };

    getInputValidity = () => {
        return {
            isIssueTypeValid: this.state.issueType.length > 0 ? true : false,
            isIssueCategoryValid: this.state.issueCategory.length > 0 ? true : false,
            isIssueDescriptionValid: this.state.issueDescription.length > 9  ? true : false,
        };
    };

    getDataToSubmit = () => {
        let data =  {
            issue_type: this.state.issueType,
            issue_category: this.state.issueCategory,
            issue_description: this.state.issueDescription
        };

        if(this.state.questionID !== '') {
            data.question_id = this.state.questionID;
        }

        if(this.props.data) {
            data._id = this.props.data._id;
        }

        return data;
    }
}

export default IssueForm;