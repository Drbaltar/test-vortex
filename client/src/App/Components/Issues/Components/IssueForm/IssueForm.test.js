import { shallow } from 'enzyme';
import React from 'react';

import IssueForm from './IssueForm';

class TestFooter extends React.Component {
    constructor(props) {
        super(props);
    }

    render() { <div />; }
}

const applicationCategories = [
    '',
    'Application Crash',
    'Entry/List Not Populating Information',
    'Incorrect Information/Description',
    'Incorrect Spelling/Terminology',
    'Server Error',
    'Request New Feature',
    'Recommend Application Change',
    'Other'
];

const questionCategories = [
    '',
    'Incorrect Question Information',
    'Outdated Information',
    'Incorrect Answer',
    'Other'
];

const initialState = {
    issueType: '',
    issueCategory: '',
    issueDescription: '',
    questionID: '',
    inputValidity: {
        isIssueTypeValid: null,
        isIssueCategoryValid: null,
        isIssueDescriptionValid: null,
        isQuestionIDValid: null
    }
};

const testData = {
    issue_type: 'Application Issue',
    issue_category: 'Application Crash',
    issue_description: 'The Application is crashing',
    question_id: ''
};

const testTitle = 'Edit Issue';

describe('IssueForm', () => {
    const wrapper = shallow(<IssueForm title={testTitle}><TestFooter /></IssueForm>);

    it('renders correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('sets the application categories to the correct entries', () => {
        expect(wrapper.instance().applicationCategories).toEqual(applicationCategories);
    });

    it('sets the question categories to the correct entries', () => {
        expect(wrapper.instance().questionCategories).toEqual(questionCategories);
    });

    it('sets the initial state to the correct values', () => {
        expect(wrapper.state()).toEqual(initialState);
    });

    describe('when there is data that is passed to the form', () => {
        const formWithData = shallow(<IssueForm data={testData}><TestFooter /></IssueForm>);

        it('sets the state values to the correct entries', () => {
            expect(formWithData.state()).toEqual(expect.objectContaining({
                issueType: testData.issue_type,
                issueCategory: testData.issue_category,
                issueDescription: testData.issue_description,
                questionID: testData.question_id
            }));
        });
    });

    it('renders the BodyCard component for the form', () => {
        expect(wrapper.find('BodyCard').exists()).toBe(true);
    });

    it('sets the BodyCard title to the passed in title prop', () => {
        expect(wrapper.find('BodyCard').prop('title')).toEqual(testTitle);
    });

    it('passes the issueType field to the BodyCard component', () => {
        expect(wrapper.find('BodyCard').childAt(0).childAt(0).prop('id')).toEqual('issueType');
    });

    it('passes the issueCategory field to the BodyCard component', () => {
        expect(wrapper.find('BodyCard').childAt(0).childAt(1).prop('id')).toEqual('issueCategory');
    });

    it('passes the issueDescription field to the BodyCard component', () => {
        expect(wrapper.find('BodyCard').childAt(0).childAt(2).prop('id')).toEqual('issueDescription');
    });

    it('passes the footer component to the BodyCard component', () => {
        expect(wrapper.find('BodyCard').childAt(1).name()).toEqual('TestFooter');
    });

    it('renders the issue type field', () => {
        expect(wrapper.find('#issueType').exists()).toBe(true);
    });

    it('passes the correct props to the issue type field', () => {
        expect(wrapper.find('#issueType').props()).toEqual({
            label: 'Issue Type',
            id: 'issueType',
            options: ['', 'Application Issue', 'Question Issue'],
            value: wrapper.state().issueType,
            inputChange: expect.any(Function),
            isValid: wrapper.state().inputValidity.isIssueTypeValid,
            errorMessage: 'You must select the issue type!'
        });
    });

    it('renders the issue category field', () => {
        expect(wrapper.find('#issueCategory').exists()).toBe(true);
    });

    it('passes the correct props to the issue category field', () => {
        expect(wrapper.find('#issueCategory').props()).toEqual({
            label: 'Issue Category',
            id: 'issueCategory',
            options: [],
            value: wrapper.state().issueCategory,
            inputChange: expect.any(Function),
            isValid: wrapper.state().inputValidity.isIssueCategoryValid,
            errorMessage: 'You must select the issue category!'
        });
    });

    describe('when the issue type field is set to `Application Issue`', () => {
        beforeAll(() => {
            wrapper.setState({ issueType: 'Application Issue' });
        });

        it('sets the issue category field options to the applicable options', () => {
            expect(wrapper.find('#issueCategory').prop('options')).toEqual(applicationCategories);
        });

        afterAll(() => {
            wrapper.setState({ issueType: '' });
        });
    });

    describe('when the issue type field is set to `Question Issue`', () => {
        beforeAll(() => {
            wrapper.setState({ issueType: 'Question Issue' });
        });

        it('sets the issue category field options to the applicable options', () => {
            expect(wrapper.find('#issueCategory').prop('options')).toEqual(questionCategories);
        });

        afterAll(() => {
            wrapper.setState({ issueType: '' });
        });
    });

    it('renders the question description field', () => {
        expect(wrapper.find('#issueDescription').exists()).toBe(true);
    });

    it('does not render the error message when errorMessage is not a passed in prop', () => {
        expect(wrapper.find('ErrorMessage').exists()).toBe(false);
    });
});