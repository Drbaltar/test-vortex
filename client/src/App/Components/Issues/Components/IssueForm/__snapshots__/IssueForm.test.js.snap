// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IssueForm renders correctly 1`] = `
<BodyCard
  title="Edit Issue"
>
  <div
    id="issue-form-body"
  >
    <SelectBox
      errorMessage="You must select the issue type!"
      id="issueType"
      inputChange={[Function]}
      isValid={null}
      label="Issue Type"
      options={
        Array [
          "",
          "Application Issue",
          "Question Issue",
        ]
      }
      value=""
    />
    <SelectBox
      errorMessage="You must select the issue category!"
      id="issueCategory"
      inputChange={[Function]}
      isValid={null}
      label="Issue Category"
      options={Array []}
      value=""
    />
    <TextField
      errorMessage="The issue description must be at least 10 characters!"
      id="issueDescription"
      inputChange={[Function]}
      isValid={null}
      label="Issue Description"
      rows="4"
      type="text"
      value=""
    />
  </div>
  <TestFooter
    clickHandler={[Function]}
  />
</BodyCard>
`;
